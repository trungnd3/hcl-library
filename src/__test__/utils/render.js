// test-utils.jsx
import React from 'react';
import { render as rtlRender } from '@testing-library/react';
import { configureStore } from '@reduxjs/toolkit';
import { Provider } from 'react-redux';
import { BrowserRouter } from 'react-router-dom';

// Import your own reducer
import authReducer from '../../app/slices/auth';
import bookReducer from '../../app/slices/book';

function render(
  ui,
  {
    preloadedState,
    store = configureStore({
      reducer: { auth: authReducer, book: bookReducer },
      preloadedState,
    }),
    ...renderOptions
  } = {}
) {
  function Wrapper({ children }) {
    return (
      <Provider store={store}>
        <BrowserRouter>{children}</BrowserRouter>
      </Provider>
    );
  }
  return rtlRender(ui, { wrapper: Wrapper, ...renderOptions });
}

// re-export everything
export * from '@testing-library/react';
// override render method
export { render };
