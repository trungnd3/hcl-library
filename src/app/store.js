import { configureStore } from '@reduxjs/toolkit';
import authReducer from './slices/auth';
import bookReducer from './slices/book';

export const store = configureStore({
  reducer: {
    auth: authReducer,
    book: bookReducer,
  },
});
