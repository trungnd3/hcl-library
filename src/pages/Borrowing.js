import { useDispatch, useSelector } from 'react-redux';

import BorrowingTable from '../components/Book/BorrowingTable';
import Centered from '../components/UI/Centered';

import { selectBorrowing, returnBook } from '../app/action-creators/book';

const Borrowing = () => {
  const borrowingBooks = useSelector(selectBorrowing);
  const dispatch = useDispatch();

  const returnBookHandler = (id) => {
    dispatch(returnBook(id));
  };

  if (!borrowingBooks.length) {
    return (
      <Centered>
        <h3>You are currently not borrowing any book.</h3>
      </Centered>
    );
  }

  return (
    <BorrowingTable books={borrowingBooks} onReturnBook={returnBookHandler} />
  );
};

export default Borrowing;
