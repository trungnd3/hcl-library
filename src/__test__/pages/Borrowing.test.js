import React from 'react';
import Borrowing from '../../pages/Borrowing';
import { Provider } from 'react-redux';
import { store } from '../../app/store';
import { prettyDOM, render, screen, fireEvent } from '../utils/render';
import userEvent from '@testing-library/user-event';
import renderer from 'react-test-renderer';

let user;

describe('Borrowing component using RTL', () => {
  beforeEach(() => {
    render(<Borrowing />);
    user = userEvent.setup();
  });

  test('render snapshot', () => {
    const tree = renderer
      .create(
        <Provider store={store}>
          <Borrowing />
        </Provider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('should display empty data', () => {
    const emptyText = screen.getByText(
      'You are currently not borrowing any book.'
    );
    expect(emptyText).toBeInTheDocument();
  });

  // test('should display borrowing table', () => {
  //   const table = screen.getByRole('table');
  //   console.log(table);
  // });
});
