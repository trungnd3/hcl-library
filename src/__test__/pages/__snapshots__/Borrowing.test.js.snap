// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Borrowing component using RTL render snapshot 1`] = `
<div
  className="mt-5 text-center"
>
  <h3>
    You are currently not borrowing any book.
  </h3>
</div>
`;
